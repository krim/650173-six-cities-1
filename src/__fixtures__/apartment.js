export default {
  city: {
    name: `Amsterdam`,
    location: {
      latitude: 51.225402,
      longitude: 6.776314,
      zoom: 13
    }
  },
  previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/1.jpg`,
  images: [
    `https://es31-server.appspot.com/six-cities/static/hotel/18.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/4.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/12.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/15.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/8.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/5.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/6.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/1.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/7.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/19.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/14.jpg`
  ],
  title: `Waterfront with extraordinary view`,
  isFavorite: false,
  isPremium: true,
  rating: 2.7,
  type: `hotel`,
  bedrooms: 2,
  maxAdults: 8,
  price: 393,
  goods: [
    `Breakfast`,
    `Washer`,
    `Laptop friendly workspace`
  ],
  host: {
    id: 25,
    name: `Angelina`,
    isPro: true,
    avatarUrl: `img/avatar-angelina.jpg`
  },
  description: `I am happy to welcome you to my apartment in the city center! Three words: location, cosy and chic!`,
  location: {
    latitude: 51.214402,
    longitude: 6.764314000000001,
    zoom: 16
  },
  id: 1
};
