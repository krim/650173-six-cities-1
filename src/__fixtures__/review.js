import user from './user';

export default {
  id: 1,
  user,
  rating: 3,
  comment: `The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.`,
  date: `2019-05-21T05:25:32.222Z`
};
