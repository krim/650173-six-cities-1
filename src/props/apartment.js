import PropTypes from 'prop-types';
import {range} from 'react-range-proptypes';

export default PropTypes.shape({
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  previewImage: PropTypes.string.isRequired,
  rating: range(0, 100).isRequired,
  price: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  bedrooms: PropTypes.number.isRequired,
  maxAdults: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isPremium: PropTypes.bool.isRequired,
  location: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
    zoom: PropTypes.number.isRequired
  }),
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    isPro: PropTypes.bool.isRequired,
    avatarUrl: PropTypes.string.isRequired
  })
});
