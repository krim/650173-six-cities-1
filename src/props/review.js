import PropTypes from 'prop-types';
import userProps from './user';

export default PropTypes.shape({
  id: PropTypes.number.isRequired,
  user: userProps,
  rating: PropTypes.number.isRequired,
  comment: PropTypes.string.isRequired,
  date: PropTypes.number.isRequired
});
