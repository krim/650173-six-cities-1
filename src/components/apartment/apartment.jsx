import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

import {apartmentProps} from '../../props';
import getRating from '../../libs/get-rating';
import withFavorite from '../../hocs/with-favorite/with-favorite';
import BookmarkButton from '../bookmark-button/bookmark-button.jsx';

const ApartmentTypes = {
  apartment: `Apartment`,
  room: `Private Room`,
  house: `House`,
  hotel: `Hotel`
};

class Apartment extends PureComponent {
  render() {
    const {
      apartment,
      onImageClick,
      onClick,
      onBookmarkClick,
      className
    } = this.props;

    return (
      <article
        className={`${this._getArticleClass()} place-card`}
      >
        { apartment.isPremium && <div className="place-card__mark"><span>Premium</span></div> }
        <div className={`${className}__image-wrapper place-card__image-wrapper`}>
          <a href="#" onClick={onImageClick}>
            <img
              className="place-card__image"
              src={apartment.previewImage}
              alt="Place image"
              {...this._getImageSize()}
            />
          </a>
        </div>

        <div className={`place-card__info${this._getCardClass()}`}>
          <div className={`place-card__price-wrapper`}>
            <div className="place-card__price">
              <b className="place-card__price-value">&euro;{apartment.price}</b>
              <span className="place-card__price-text">&#47;&nbsp;night</span>
            </div>
            <BookmarkButton
              className={`place-card`}
              height={`19`}
              width={`18`}
              isFavorite={apartment.isFavorite}
              onBookmarkClick={onBookmarkClick}
            />
          </div>
          <div className="place-card__rating rating">
            <div className="place-card__stars rating__stars">
              <span style={{width: getRating(apartment.rating)}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <h2 className="place-card__name">
            <a href="#" onClick={onClick}>{apartment.title}</a>
          </h2>
          <p className="place-card__type">{ApartmentTypes[apartment.type]}</p>
        </div>
      </article>
    );
  }

  _getCardClass() {
    return this._isFavorite() ? ` favorites__card-info` : ``;
  }

  _getArticleClass() {
    return this._isFavorite() ? ` favorites__card` : `cities__place-card`;
  }

  _getImageSize() {
    return this._isFavorite() ? {width: `150`, height: `110`} : {width: `260`, height: `200`};
  }

  _isFavorite() {
    return this.props.className === `favorites`;
  }
}

Apartment.propTypes = {
  apartment: apartmentProps,
  onClick: PropTypes.func,
  onImageClick: PropTypes.func,
  onBookmarkClick: PropTypes.func,
  className: PropTypes.string.isRequired
};

export {Apartment, ApartmentTypes};
export default withFavorite(Apartment);

