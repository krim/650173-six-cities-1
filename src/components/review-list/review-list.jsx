import React, {Component} from 'react';

import {reviewProps} from '../../props';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import Review from '../review/review.jsx';
import ReviewForm from '../review-form/review-form.jsx';
import {Operation} from '../../reducer/data/data';
import {getSortedReviews} from '../../reducer/data/selectors';

class ReviewList extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.loadReviews(this.props.apartmentId);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.apartmentId !== this.props.apartmentId) {
      this.props.loadReviews(this.props.apartmentId);
    }
  }

  render() {
    const {reviews, apartmentId, isUserAuthorized} = this.props;

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
        <ul className="reviews__list">
          {
            reviews.map((review) => {
              return (
                <li className="reviews__item" key={`review-${review.id}`}>
                  <Review review={review} />
                </li>
              );
            })
          }
        </ul>

        {isUserAuthorized && <ReviewForm apartmentId={apartmentId}/>}
      </section>
    );
  }
}

ReviewList.propTypes = {
  apartmentId: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(reviewProps),
  loadReviews: PropTypes.func.isRequired,
  isUserAuthorized: PropTypes.bool.isRequired
};

const mapStateToProps = (state) => {
  return {
    reviews: getSortedReviews(state)
  };
};

const mapDispatchToProps = (dispatch) => ({
  loadReviews: (id) => dispatch(Operation.loadReviews(id))
});

export {ReviewList};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewList);
