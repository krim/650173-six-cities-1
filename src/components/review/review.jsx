import React, {PureComponent} from 'react';

import {reviewProps} from '../../props';
import getRating from '../../libs/get-rating';

class Review extends PureComponent {
  render() {
    const {review} = this.props;
    const {user} = review;

    return (
      <React.Fragment>
        <div className="reviews__user user">
          <div className="reviews__avatar-wrapper user__avatar-wrapper">
            <img
              className="reviews__avatar user__avatar"
              src={user.avatarUrl}
              width="54"
              height="54"
              alt="Reviews avatar"
            />
          </div>
          <span className="reviews__user-name">
            { user.name }
          </span>
        </div>
        <div className="reviews__info">
          <div className="reviews__rating rating">
            <div className="reviews__stars rating__stars">
              <span style={{width: getRating(review.rating)}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <p className="reviews__text">
            {review.comment}
          </p>
          <time className="reviews__time" dateTime={this._getFormatedDateTime()}>{this._getFormatedDate()}</time>
        </div>
      </React.Fragment>
    );
  }

  _getFormatedDateTime() {
    const date = this._getReviewDate();
    const options = {year: `numeric`, month: `2-digit`, day: `2-digit`};
    return new Date(date).toLocaleString(`se-SE`, options);
  }

  _getFormatedDate() {
    const date = this._getReviewDate();
    const options = {year: `numeric`, month: `long`};
    return new Date(date).toLocaleString(`en-US`, options);
  }

  _getReviewDate() {
    return Date.parse(this.props.review.date);
  }
}

Review.propTypes = {
  review: reviewProps
};

export default Review;

