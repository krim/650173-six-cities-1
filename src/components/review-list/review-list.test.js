import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';

import {ReviewList} from './review-list';
import apartment from '../../__fixtures__/apartment';
import review from '../../__fixtures__/review';

const reviews = [review];
const middlewares = [];
const mockStore = configureMockStore(middlewares);
const initialState = {};
const store = mockStore(initialState);

describe(`ReviewList`, () => {
  it(`renders component correctly`, () => {
    const tree = renderer.create(
        <Provider store={store}>
          <ReviewList
            apartmentId={apartment.id}
            reviews={reviews}
            loadReviews={jest.fn()}
          />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
