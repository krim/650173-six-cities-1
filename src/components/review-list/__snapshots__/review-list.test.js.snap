// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewList when user is authorized renders component correctly 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      1
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://es31-server.appspot.com/six-cities/static/avatar/4.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Zak
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-21"
        >
          May 2019
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    method="post"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="5-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="5"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="4-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="4"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="3-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="3"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not bad"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="2-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="2"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="1-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="1"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-stars"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      id="review"
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
      value=""
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;

exports[`ReviewList when user is not authorized renders component correctly 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      1
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://es31-server.appspot.com/six-cities/static/avatar/4.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Zak
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-21"
        >
          May 2019
        </time>
      </div>
    </li>
  </ul>
</section>
`;
